src/Data/HashMap/Linear.hs

```diff
@@ -1,73 +1,239 @@
+--     and this could be done away with
+--   * We don't resize to make the array smaller when we have too much space
+--   * Make the lookup use smart search by probing from the mean PSL:
+--     probe at the mean, mean-1, mean+1, mean-2, mean+2, ...
+--   * Other functions from Data.Map
 
 module Data.HashMap.Linear
-  (
+  ( singleton,
+    alter,
+    insert,
+    delete,
+    size,
+    member,
+    lookup
   )
 where
 
-
-{-
-
 import Data.Hashable
+import Data.Unrestricted.Linear
 import Data.Array.Mutable.Linear
-
+import Prelude.Linear hiding (read, lookup, (+))
+import Prelude ((+))
+import qualified Prelude
+import qualified Unsafe.Linear as Unsafe
 
 -- # Core Data Types
 --------------------------------------------------
 
 -- | Robin values are triples of the key, value and PSL
 -- (the probe sequence length).
-type RobinVal k v = (k,v,Int)
+type RobinVal k v = (k,v,PSL)
 
-data HashMap k v where
-  -- | HashMap size array-of-robin-values
-  HashMap :: Int -> Array (RobinVal k v) #-> HashMap k v
+-- | A probe sequence length
+newtype PSL = PSL Int
+  deriving (Prelude.Eq, Prelude.Ord, Prelude.Num)
+
+-- | An array of Robin values
+type RobinArr k v = Array (RobinVal k v)
 
+-- | At minimum, we need to store hashable
+-- and identifiable keys
+type Keyable k = (Eq k, Hashable k)
+
+-- | A mutable hashmap with a linear API
+data HashMap k v where
+  -- | HashMap (size,count) array-of-robin-values
+  -- The size is the amount of memory the array takes up.
+  -- The count is the number of stored mappings.
+  -- Our sparseness invariant: count*3 <= size.
+  HashMap :: (Int, Int) -> RobinArr k v #-> HashMap k v
+
+-- INVARIANTS:
+--   * Cells are empty iff the PSL is -1.
+--   * We are sparse: count*3 <= size, with correct values of count and size
+--   * Each (RobinVal k v) has the correct PSL
+
+-- | The results of searching for where to insert a key
+data RobinQuery k where
+  -- | A key, PSL pair to insert at an index, that has not
+  -- yet been touched. Invariant: the PSL must be -1 at that index.
+  IndexToInsert :: (k, PSL) -> Int -> RobinQuery k
+  -- A key, PSL pair to update the value at an index.
+  IndexToUpdate :: (k, PSL) -> Int -> RobinQuery k
+  -- | A key, PSL pair for a swap at an index.
+  -- The swapped-out pair will then need to be inserted downstream.
+  IndexToSwap :: (k, PSL) -> Int -> RobinQuery k
 
 -- # Construction and Modification
 --------------------------------------------------
 
-singleton :: Hashable k =>
+-- | Run a computation using a singleton hashmap
+singleton :: Keyable k =>
   (k,v) -> (HashMap k v #-> Unrestricted b) -> Unrestricted b
-singleton = undefined
-
-alter ::  Hashable k => (Maybe a -> Maybe a) -> k -> HashMap k v #-> HashMap k v
-alter = undefined
-
-insert :: Hashable k => HashMap k v #-> k -> v -> HashMap k v
-insert hmap k v = alter (\_ -> Just v) k hmap
+singleton (k :: k, v :: v) (f :: HashMap k v #-> Unrestricted b) =
```

Comment:
By the way, since you are not using `ScopedTypeVariables`, these type annotation don't have the meaning you think they have.

Reply:
I am using `ScopedTypeVariables`.

Reply:
I'm leaving it as a puzzle to you to solve, how you are not making use of it enough in this particular case.

It shouldn't prevent you from merging this, of course.